DataPipes and DataFrames integration

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64309 [WIP] Add depth tracking
* **#60840 DataPipes and DataFrames integration**

Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

VitalyFedyunin committed:
[WIP] Example of DataPipes and DataFrames integration

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"




[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/dataframes.py:
```diff
@@ -0,0 +1,416 @@
+        aggregate = []
+        for item in self.source_datapipe:
+            aggregate.append(self._as_list(item))
+            if len(aggregate) == self.dataframe_size:
```
`len(aggregate)` can be larger than `dataframe_size` here. We need to yield first `dataframe_size` of items, and keep the rest in the buffer

ejguan commented on torch/utils/data/datapipes/iter/dataframes.py:
```diff
@@ -0,0 +1,416 @@
+@functional_datapipe('dataframes_as_tuples')
+class DataFramesAsTuplesPipe(IterDataPipe):
+    def __init__(self, source_datapipe):
+        self.source_datapipe = source_datapipe
```
Add assert that `isinstance(source_datapipe, DFIterDataPipe)`

ejguan commented on torch/utils/data/datapipes/iter/dataframes.py:
```diff
@@ -0,0 +1,416 @@
+
+
+@functional_datapipe('dataframes_concat', is_df=True)
+class ShuffleDataFramesPipe(DFIterDataPipe):
```
nit: ConcatDataFramesPipe

VitalyFedyunin replied:
done

ejguan commented on torch/utils/data/datapipes/iter/dataframes.py:
```diff
@@ -0,0 +1,416 @@
+
+
+@functional_datapipe('dataframes_filter', is_df=True)
+class ShuffleDataFramesPipe(DFIterDataPipe):
```
nit: FilterDataFramesPipe

VitalyFedyunin replied:
done

ejguan commented on torch/utils/data/datapipes/iter/dataframes.py:
```diff
@@ -0,0 +1,416 @@
+try:
+    import pandas
+    WITH_PANDAS = True
+finally:
```
```suggestion
except ImportError:
```

VitalyFedyunin replied:
done

ejguan reviewed (changes requested):
A few comments for this PR.
And, do you think we should also align the APIs like `dataframe_shuffle` with `shuffle`? The stack of PRs has added like `unbatch` to the API, should we keep them same?

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

wenleix commented on torch/utils/data/__init__.py:
```diff
@@ -2,8 +2,17 @@
+     Dataset as MapDataPipe,
+     DFIterDataPipe as DFIterDataPipe,
+     IterableDataset,
+     IterableDataset as IterDataPipe,
```
nit: why `import IterableDataset` and `import IterableDataset as IterDataPipe`?

VitalyFedyunin replied:
cleaned

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

wenleix commented on torch/utils/data/dataframes_pipes.ipynb:
```diff
@@ -0,0 +1,523 @@
+   ],
+   "source": [
+    "dp = get_dataframes_pipe(dataframe_size = 3)\n",
+    "dp['y'] = dp.i * 100 + dp.j - 2.7\n",
```
can i use `dp['i']` here as well ? e.g. 
```
dp['y'] = dp['i'] * 100 + dp['j'] - 2.7
```

VitalyFedyunin replied:
yes

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

wenleix commented on torch/utils/data/datapipes/iter/dataframes.py:
```diff
@@ -0,0 +1,434 @@
+                size = len(df.index)
+            for i in range(len(df.index)):
+                all_buffer.append(df[i:i + 1])
+                filter_res.append(self.filter_fn(df.iloc[i]))
```
it looks like `df.iloc[i]` returns the ith series (i.e. Column), while `df.iloc[[i]]` returns the ith row (confusing Pandas syntax, lol) . https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[WIP] Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Example of DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/dataframe/dataframes.py:
```diff
@@ -0,0 +1,294 @@
+
+
+#  TODO(VitalyFedyunin): Extract this list from the DFIterDataPipe registred functions
+DATAPIPES_OPS = ['dataframes_as_tuples', 'groupby', 'dataframes_filter', 'map', 'to_datapipe',
+                 'shuffle', 'concat', 'batch', 'dataframes_per_row', 'dataframes_concat', 'dataframes_shuffle']
```
Are we exposing `dataframes_as_tuples`, `dataframes_per_row`, ... these API to users?

VitalyFedyunin replied:
I will underscore them =)

ejguan commented on torch/utils/data/datapipes/dataframe/datapipes.py:
```diff
@@ -0,0 +1,137 @@
+                yield record
+
+
+@functional_datapipe('dataframes_per_row', enable_df_api_tracing=True)
```
If these APIs are not user-facing, can we remove the functional API?

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -82,12 +87,26 @@ def register_function(cls, function_name, function):
-            return cls(source_dp, *args, **kwargs)
+            result_pipe = cls(source_dp, *args, **kwargs)
+            if isinstance(result_pipe, Dataset):
+                if enable_df_api_tracing or isinstance(source_dp, DFIterDataPipe) or getattr(result_pipe, '_dp_cast_to_df', False):
+                    if function_name not in UNTRACABLE_DATAFRAME_PIPES:
+                        result_pipe = result_pipe.trace_as_dataframe()
+
+                if getattr(result_pipe, '_dp_nesting_depth', None) is None:
+                    result_pipe._dp_nesting_depth = getattr(source_dp, '_dp_nesting_depth', None)
+
+                if getattr(result_pipe, '_dp_contains_dataframe', None) is None:
+                    result_pipe._dp_contains_dataframe = getattr(source_dp, '_dp_contains_dataframe', None)
```
Could we add comments toward attribute of `_dp_cast_to_df`, `_dp_nesting_depth` and `_dp_contains_dataframe`?

ejguan commented on test/test_datapipe.py:
```diff
@@ -343,6 +350,56 @@ def test_demux_mux_datapipe(self):
         self.assertEqual(list(range(10)), list(n))
 
 
+class TestDataFramesPipes(TestCase):
```
Is there way to test whether the Capture object is correct?

VitalyFedyunin replied:
For users they are printable (but with dynamic names), for code, execute and verify results probably the best approach.

ejguan reviewed (commented):
I think it looks reasonable to me with a few questions.
BTW, are we going to really land it into core? Can we move it into torchdata? Is there any dependency issue

ejguan commented:
As long as we make all methods `Capture` to different DFIterDataPipe rather than embedded into normal IterDataPipe, I think we can untackle the dependency issue.

VitalyFedyunin commented:
> As long as we make all methods `Capture` to different DFIterDataPipe rather than embedded into normal IterDataPipe, I think we can untackle the dependency issue.

My think was to make DFPipes first class citizen to make sure all further development take them into account. As proposal we land them to core now, and work on decoupling after discovering all connection scenarios (there are few things remains to explore like tracking nesting depth properly, and behavior of standard pipes like grouping etc.)

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/dataframe/dataframes.py:
```diff
@@ -0,0 +1,294 @@
+            dp = self.unbatch(unbatch_level)._dataframes_per_row()
+        else:
+            dp = self._dataframes_per_row()
+        dp = dp.as_datapipe().groupby(group_key_fn, buffer_size=buffer_size, group_size=group_size,
+                                      guaranteed_group_size=guaranteed_group_size, drop_remaining=drop_remaining)
+        dp._dp_contains_dataframe = True
+        dp._dp_nesting_depth = 1
```
This is going to assign `_dp_nesting_depth` to 1

VitalyFedyunin replied:
Moved all nesting to the next PR, as they still requires some cleanups.

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -65,7 +66,11 @@ def __init__(self,
         self.drop_last = drop_last
         self.length = None
-        self.wrapper_class = DataChunk
+        source_depth = getattr(datapipe, '_dp_nesting_depth', 0)
+        if source_depth is None:
+            source_depth = 0
+        self._dp_nesting_depth = source_depth + 1
```
We then increment `_dp_nesting_depth` from the normal `groupby`. Is this part really required?

VitalyFedyunin replied:
Moved all nesting to the next PR, as they still requires some cleanups.

ejguan reviewed (approved):
> My think was to make DFPipes first class citizen to make sure all further development take them into account. As proposal we land them to core now, and work on decoupling after discovering all connection scenarios 

That sounds great. We may need Wenlei to go over it again.

> (there are few things remains to explore like tracking nesting depth properly, and behavior of standard pipes like grouping etc.)

This is the part I am still not sure. Please see the inline comments.

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

wenleix commented on torch/utils/data/datapipes/dataframe/dataframes.py:
```diff
@@ -0,0 +1,292 @@
+        return iter(self._dataframes_as_tuples())
+
+    def batch(self, batch_size=10):
+        dp = self._dataframes_per_row()._dataframes_concat(batch_size)
```
Curious on this: After `_dataframes_per_row` , the `DataPipe` will return one row a t a time (so it flatten the DataFrame`). 

But`_dataframes_concat`  will concat multiple DataFrames together?

wenleix reviewed (approved):
Thanks!

One high level question is whether there is any way to quickly identify a given `DataPipe` is a per-DF or per-row iterable? -- for example, sometimes it's not obvious whether the `source_datapipe` is a per-DF or per-row DataPipe without looking at the callsite, etc.

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

VitalyFedyunin committed:
Update on " DataPipes and DataFrames integration"


Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).
